pkg/asset/agent/joiner/clusterinfo.go

```diff
@@ -0,0 +1,232 @@
+	ImageDigestSources            []types.ImageDigestSource
+	DeprecatedImageContentSources []types.ImageContentSource
+	PlatformType                  hiveext.PlatformType
+	SSHKey                        string
+}
+
+var _ asset.WritableAsset = (*ClusterInfo)(nil)
+
+// Name returns the human-friendly name of the asset.
+func (ci *ClusterInfo) Name() string {
+	return "Agent Installer ClusterInfo"
+}
+
+// Dependencies returns all of the dependencies directly needed to generate
+// the asset.
+func (*ClusterInfo) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		&workflow.AgentWorkflow{},
+		&AddNodesConfig{},
+	}
+}
+
+// Generate generates the ClusterInfo.
+func (ci *ClusterInfo) Generate(dependencies asset.Parents) error {
+	agentWorkflow := &workflow.AgentWorkflow{}
+	addNodesConfig := &AddNodesConfig{}
+	dependencies.Get(agentWorkflow, addNodesConfig)
+
+	if agentWorkflow.Workflow != workflow.AgentWorkflowTypeAddNodes {
+		return nil
+	}
+
+	config, err := ci.getRestConfig(addNodesConfig.Params.Kubeconfig)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveAPIDNSName(config)
+	if err != nil {
+		return err
+	}
+
+	clientset, err := configclient.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveClusterData(clientset)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveProxy(clientset)
+	if err != nil {
+		return err
+	}
+
+	k8sclientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	err = ci.retrievePullSecret(k8sclientset)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveUserTrustBundle(k8sclientset)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveArchitecture(k8sclientset)
+	if err != nil {
+		return err
+	}
+	err = ci.retrieveInstallConfigData(k8sclientset)
+	if err != nil {
+		return err
+	}
+
+	ci.Namespace = "cluster0"
+
+	return nil
+}
+
+func (ci *ClusterInfo) getRestConfig(kubeconfig string) (*rest.Config, error) {
+	var err error
+	var config *rest.Config
+
+	if kubeconfig != "" {
+		config, err = clientcmd.BuildConfigFromFlags("", kubeconfig)
+	} else {
+		config, err = rest.InClusterConfig()
+	}
+
+	return config, err
+}
+
+func (ci *ClusterInfo) retrieveClusterData(clientset *configclient.Clientset) error {
+	cv, err := clientset.ConfigV1().ClusterVersions().Get(context.Background(), "version", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	ci.ClusterID = string(cv.Spec.ClusterID)
+	ci.ReleaseImage = cv.Status.History[0].Image
+	ci.Version = cv.Status.History[0].Version
+
+	return nil
+}
+
+func (ci *ClusterInfo) retrieveProxy(clientset *configclient.Clientset) error {
+	proxy, err := clientset.ConfigV1().Proxies().Get(context.Background(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	ci.Proxy = &types.Proxy{
+		HTTPProxy:  proxy.Spec.HTTPProxy,
+		HTTPSProxy: proxy.Spec.HTTPSProxy,
+		NoProxy:    proxy.Spec.NoProxy,
+	}
+
+	return nil
+}
+
+func (ci *ClusterInfo) retrieveAPIDNSName(config *rest.Config) error {
```

Comment:
Could we create a unit test for this function? That may help people see where in the kubeconfig file the APIDNSName is coming from.

Reply:
I was planning to add a complete unit test for the whole ClusterInfo asset, added it